crates/derive/src/types/batch.rs

```diff
@@ -0,0 +1,42 @@
+//! This module contains the enumerable [Batch].
+
+use super::batch_type::BatchType;
+use super::single_batch::SingleBatch;
+use crate::types::errors::DecodeError;
+
+use alloy_rlp::Decodable;
+
+// TODO: replace this with a span batch
+/// Span Batch.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SpanBatch {}
+
+/// A Batch.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Batch {
+    /// A single batch
+    Single(SingleBatch),
+    /// Span Batches
+    Span(SpanBatch),
+}
+
+impl TryFrom<&[u8]> for Batch {
```

Comment:
Can we make this a `decode` function in an `impl Batch` block?

When we have a reader (`&mut &[u8]`), we want to update the ptr as the data is read from it. With this we have to pass a reference to the actual data rather than the ptr, so our reader wouldn't get updated.

```rs
let buf = [0, 1, 2, 3, 4, ...];
let reader = &mut &buf;
let batch = Batch.try_from(reader).unwrap();
// reader was not advanced by the batch data read here
```